File: Data_Structures/python/Linked_lists/deleteMiddleNode.py
class Node:

    def __init__(self, data):

        self.data = data
        self.next = None

class LinkedList:

    def __init__(self):
        self.head = None

    def addToList(self, data):

        newNode = Node(data)
        if self.head is None:
            self.head = newNode
            return

        last = self.head

        while last.next:
            last = last.next

        last.next = newNode

    # Returns the list in string format
    def __str__(self):

        linkedListStr = ""
        temp = self.head

        while temp:
            linkedListStr += str(temp.data) + "->"
            temp = temp.next

        return linkedListStr + "NULL"

    # Method deletes middle node
    def deleteMid(self):

        # Base cases
        if (self.head is None or self.head.next is None):
            return

        # Initialize slow and fast pointers
        # to reach middle of linked list
        slow_Ptr = self.head
        fast_Ptr = self.head

        # Find the middle and previous of middle
        prev = None

        # To store previous of slow pointer
        while (fast_Ptr is not None and fast_Ptr.next is not None):
            fast_Ptr = fast_Ptr.next.next
            prev = slow_Ptr
            slow_Ptr = slow_Ptr.next

        # Delete the middle node
        prev.next = slow_Ptr.next

# Driver code
linkedList = LinkedList()

linkedList.addToList(1)
linkedList.addToList(2)
linkedList.addToList(3)
linkedList.addToList(4)

print("Given Linked List")
print(linkedList)

linkedList.deleteMid()

print("Linked List after deletion of middle")
print(linkedList)

File: Data_Structures/C++/Linked_lists/deleteMiddleNode.cpp
#include <iostream>

using namespace std;

class Node {
    int data;
    Node *next;
}

int     countNodes(Node *head)
{
    int count = 0;
    while (head)
    {
        count++;
        head = head->next;
    }
    return count;
}

void    deleteMiddleNode(Node *head)
{
    Node *it = head;
    if (head == NULL)
        return NULL;
    if (head->next == NULL) {
        delete head;
        return NULL;
    }
    int mid = countNodes(head) / 2;
    while (mid-- > 1)
        it = it->next;
    head->next = head->next->next;
    return it;
}

int     main(void)
{
    return (0);
}
